crates/sui-indexer/src/handlers/tx_processor.rs

```diff
@@ -0,0 +1,320 @@
+}
+
+impl TxChangesProcessor {
+    pub fn new(objects: &[&Object], metrics: IndexerMetrics) -> Self {
+        let mut object_cache = InMemObjectCache::new();
+        for obj in objects {
+            object_cache.insert_object(<&Object>::clone(obj).clone());
+        }
+        Self {
+            object_cache,
+            metrics,
+        }
+    }
+
+    pub(crate) async fn get_changes(
+        &self,
+        tx: &TransactionData,
+        effects: &TransactionEffects,
+        tx_digest: &TransactionDigest,
+    ) -> IndexerResult<(
+        Vec<sui_json_rpc_types::BalanceChange>,
+        Vec<IndexedObjectChange>,
+    )> {
+        let _timer = self
+            .metrics
+            .indexing_tx_object_changes_latency
+            .start_timer();
+        let object_change: Vec<_> = get_object_changes(
+            self,
+            tx.sender(),
+            effects.modified_at_versions(),
+            effects.all_changed_objects(),
+            effects.all_removed_objects(),
+        )
+        .await?
+        .into_iter()
+        .map(IndexedObjectChange::from)
+        .collect();
+        let balance_change = get_balance_changes_from_effect(
+            self,
+            effects,
+            tx.input_objects().unwrap_or_else(|e| {
+                panic!(
+                    "Checkpointed tx {:?} has inavlid input objects: {e}",
+                    tx_digest,
+                )
+            }),
+            None,
+        )
+        .await?;
+        Ok((balance_change, object_change))
+    }
+}
+
+#[async_trait]
+impl ObjectProvider for TxChangesProcessor {
+    type Error = IndexerError;
+
+    async fn get_object(
+        &self,
+        id: &ObjectID,
+        version: &SequenceNumber,
+    ) -> Result<Object, Self::Error> {
+        let object = self
+            .object_cache
+            .get(id, Some(version))
+            .as_ref()
+            .map(|o| <&Object>::clone(o).clone());
+        if let Some(o) = object {
+            self.metrics.indexing_get_object_in_mem_hit.inc();
+            return Ok(o);
+        }
+
+        panic!(
+            "Object {} is not found in TxChangesProcessor as an ObjectProvider (fn get_object)",
+            id
+        );
+    }
+
+    async fn find_object_lt_or_eq_version(
+        &self,
+        id: &ObjectID,
+        version: &SequenceNumber,
+    ) -> Result<Option<Object>, Self::Error> {
+        // First look up the exact version in object_cache.
+        let object = self
+            .object_cache
+            .get(id, Some(version))
+            .as_ref()
+            .map(|o| <&Object>::clone(o).clone());
+        if let Some(o) = object {
+            self.metrics.indexing_get_object_in_mem_hit.inc();
+            return Ok(Some(o));
+        }
+
+        // Second look up the latest version in object_cache. This may be
+        // called when the object is deleted hence the version at deletion
+        // is given.
+        let object = self
```

Comment:
if the second latest version is 90 then the input version must be 90. Otherwise the tx will fail input checks and will not reach indexer. Or I'm not following what you are referring to??

Reply:
I meant if the "second latest version" is 90 and input version arg is 91, 
- there is no exact match of 90
- the latest version is 100, which is greater than 91
so this block of codes will panic?

Reply:
resolved offline, i will rename this function to a more readable name in a follow up PR